test/src/unit-resource-pool.cc

```diff
@@ -72,4 +74,102 @@ TEST_CASE("Buffer: Test resource pool", "[resource-pool]") {
     REQUIRE(found[1]);
     REQUIRE(found[2]);
   }
+}
+
+TEST_CASE("Buffer: Test blocking resource pool", "[resource-pool]") {
+  std::thread t1;
+
+  BlockingResourcePool<int> pool(3);
+
+  ResourceGuard r1(pool);
+  ResourceGuard r2(pool);
+  r1.get() = 7;
+  r2.get() = 8;
+
+  {
+    // Get another resouce to reach maximum pool capacity
+    ResourceGuard r3(pool);
+    r3.get() = 9;
+
+    // Request a resource when pool is full, the thread should block.
+    t1 = std::thread([&] {
+      ResourceGuard r4(pool);
+      r4.get() = 10;
+    });
+
+    // Validate we can get access to the first 3 resources.
+    REQUIRE(r1.get() == 7);
+    REQUIRE(r2.get() == 8);
+    REQUIRE(r3.get() == 9);
+
+    // take r3 resource out of scope to release it
+  }
+
+  t1.join();
+
+  ResourceGuard r4(pool);
+
+  REQUIRE(r1.get() == 7);
+  REQUIRE(r2.get() == 8);
+  // check the old resource is gone, and one of the new ones is there
+  REQUIRE(r4.get() == 10);
+}
+
+TEST_CASE(
```

Comment:
@eric-hughes-tiledb I wrote this test to test that the deadlock you explained in your review was fixed, but since it uses `sleep_for` for pushing towards a certain order of events we may not want to commit it to avoid random failures in the CI due to scheduling.

Reply:
You can get deterministic order of events in tests like these with the following pattern:
* Create two threads to exercise the unit under test. Control these two threads from the test thread.
* Declare synchronization primitives in the test body and reference them in the body of the exercising threads.
  * `mutex` is adequate, though clunky. To be reliable, you need a separate mutex for each blocking point. And you need a way to call back into the control thread.
  * `barrier` and `latch` are in C++20 (and Boost) and work more easily
* Control order of execution by forcing blocking in the control thread.
  * In an exercise thread, `m1.lock()` will block if `m1` is locked in the control thread.
* With a bit of care you can ensure that at most one thread is active at a time, with the other blocked.
  * lock-in-2-then-unlock-in-1 (and vice-versa) is a standard pattern

You can remove the test from running by default with the Catch2 tag `[.]`. That would allow you to leave the test in source, with documented inadequacies, for manual testing. It probably ought to be tested thoroughly, but barring a `barrier`, it's probably not worth doing.

Reply:
Thanks for the detailed how-to, that's very interesting. Took a note and as soon as we move to C++20, I'll try out `barrier`.